src/utils/tcti.rs

```diff
@@ -0,0 +1,451 @@
+            port: Some(1234),
+            host: Some(ServerAddress::Ip(IpAddr::V4(std::net::Ipv4Addr::new(
+                168, 0, 0, 1
+            ))))
+        })
+    );
+
+    let tcti = Tcti::from_str("mssim").unwrap();
+    assert_eq!(
+        tcti,
+        Tcti::Mssim(MssimConfig {
+            port: None,
+            host: None
+        })
+    );
+
+    let tcti = Tcti::from_str("device:/try/this/path").unwrap();
+    assert_eq!(
+        tcti,
+        Tcti::Device(DeviceConfig {
+            path: Some(PathBuf::from("/try/this/path")),
+        })
+    );
+
+    let tcti = Tcti::from_str("device").unwrap();
+    assert_eq!(tcti, Tcti::Device(DeviceConfig { path: None }));
+
+    let tcti = Tcti::from_str("tabrmd:bus_name=some.bus.Name2,bus_type=session").unwrap();
+    assert_eq!(
+        tcti,
+        Tcti::Tabrmd(TabrmdConfig {
+            bus_name: Some(String::from("some.bus.Name2")),
+            bus_type: Some(BusType::Session)
+        })
+    );
+
+    let tcti = Tcti::from_str("tabrmd").unwrap();
+    assert_eq!(
+        tcti,
+        Tcti::Tabrmd(TabrmdConfig {
+            bus_name: None,
+            bus_type: None
+        })
+    );
+}
+
+/// Configuration for a Device TCTI context
+///
+/// The default configuration uses the library default of
+/// `/dev/tpm0`.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct DeviceConfig {
+    /// Path to the device node to connect to
+    ///
+    /// If set to `None`, the default location is used
+    path: Option<PathBuf>,
+}
+
+impl FromStr for DeviceConfig {
+    type Err = Error;
+
+    fn from_str(config_str: &str) -> Result<Self, Self::Err> {
+        if config_str.is_empty() {
+            return Ok(Default::default());
+        }
+
+        Ok(DeviceConfig {
+            path: Some(PathBuf::from(config_str)),
+        })
+    }
+}
+
+#[test]
+fn validate_from_str_device_config() {
+    let config = DeviceConfig::from_str("").unwrap();
+    assert!(config.path.is_none());
+
+    let config = DeviceConfig::from_str("/dev/tpm0").unwrap();
+    assert_eq!(config.path.unwrap(), PathBuf::from("/dev/tpm0"));
+}
+
+/// Configuration for an Mssim TCTI context
+///
+/// The default configuration will point to `localhost:2321`
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct MssimConfig {
+    /// Address of the server to connect to
+    ///
+    /// If set to `None`, `localhost` is used
+    host: Option<ServerAddress>,
+    /// Port used by the server at the address given in `host`
+    ///
+    /// If set to `None`, `2321` is used
+    port: Option<u16>,
+}
+
+impl FromStr for MssimConfig {
+    type Err = Error;
+
+    fn from_str(config_str: &str) -> Result<Self, Self::Err> {
+        if config_str.is_empty() {
+            return Ok(Default::default());
+        }
+        let host_pattern = Regex::new(r"(,|^)host=(.*?)(,|$)").unwrap(); // should not fail
+        let host = match host_pattern.captures(config_str) {
+            None => None,
```

Comment:
Just an idea: instead of having `Option` in the configuration, we could always put the default value instead of `None`.

Reply:
We could, but what if a user wants to set one value to something and the other leave it as the default? You can't implement `Default` on `u16`, for example, so you have to set them both manually when creating the object. I guess you could initialise with `Default::default()` and then change the one you want to change, but it feels nicer to say "no, I only want to set this one, the other doesn't matter"

Reply:
I think that if you implement `Default` on the different config structures, then an user can do:

```rust
let mssim_config = MssimConfig {
    port: 1234,
    ..Default::default()
};
```

but as I said, this looks fine as it is now and is not necessary.

Reply:
Nice, I didn't know that works, thought `Default::default()` would set the type-specific default values, not those defined for the container type.